subsys/bluetooth/host/l2cap.c

```diff
@@ -35,7 +35,8 @@
 #define L2CAP_LE_MAX_CREDITS		(CONFIG_BT_RX_BUF_COUNT - 1)
 #endif
 
-#define L2CAP_LE_CREDITS_THRESHOLD(_creds) (_creds / 2)
```

Comment:
With _creds / N, you can only set the threshold to 100% (send credit on every packet received), 50%, 33%, 25%, ...
If I would change the threshold from the default 50%, I would choose something higher to improve bandwidth, but not 100%, since this would generate too much credit traffic.
Some platforms don’t support efficient division.

For these reasons, I suggest that 

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	(_creds / CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD)
```
is changed to:

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	(_creds - (_creds >> CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD)
```

I.e. you subtract 2^-(N) of 100%, meaning that for N=0,1,2,3, you get a threshold of 100%, 50%, 75%, 87.5%.

Since most platforms do mult efficiently, it is probably better to do this:

```
#define L2CAP_LE_CREDITS_THRESHOLD(_creds) \
	((_creds * CONFIG_BT_L2CAP_RX_CREDIT_THRESHOLD) >> 8)
```

This would give e.g. these thresholds:
0	0.0%
64	25.0%
128	50.0%
192	75.0%
256	100.0%

Maybe it would be simpler and more flexible to have a callback to get the threshold from the channel? By having a CONFIG, you decide for all channels globally, which may not be what you want. There are also many configs already :-)

Reply:
Yep, Ive been debating to add a callback to read the credit threshold, perhaps with that we don't need to add set_credits function and just leave it up for the channel owner. I will play with this idea and let you know.